### PR TITLE
Feat initial rag implementation

### PR DESCRIPTION
**Pull request by lennertjansen**

### CODE REVIEW

File: backend/app/core/logging.py [hunk 1]
```diff
@@ -25,8 +25,8 @@ def configure_logger(name: str):
         # Check if the logger already has handlers to avoid duplicates
         if not logger.handlers:
             if settings.LOG_APPINSIGHTS:
-                ai_handler = AzureLogHandler(connection_string=settings.APPLICATIONINSIGHTS_CONNECTION_STRING)
-                logger.addHandler(ai_handler)
+                app_insights_handler = AzureLogHandler(connection_string=settings.APPLICATIONINSIGHTS_CONNECTION_STRING)
```
Comment (orhanrauf) [inline backend/app/core/logging.py:28]: Good change, but should be caught by static code analyzer later in refactor batch

File: backend/app/core/task_definition_retriever.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                    metric=similarity_metric,
+                    spec=PodSpec(environment=environment, pod_type=pod_type),
+                    timeout=timeout_threshold,
+                )
+                logger.info(
+                    f"Successfully created index '{index_name}'. Index stats: {self.client.describe_index(name=index_name)}"
+                )
+            except TimeoutError as exception:
+                logger.error(f"Timeout occurred while initializing the Pinecone index '{index_name}'.", exc_info=True)
+                raise
```
Comment (orhanrauf) [inline backend/app/core/task_definition_retriever.py:73]: Log-and-throw antipattern; only a single class both responsible and capable should handle exceptions

File: backend/app/core/task_definition_retriever.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    Service class for managing Pinecone vector database operations.
+
+    Attributes:
+        client (Pinecone): An instance of the Pinecone client configured with the provided API key.
+    """
+
+    def __init__(self, api_key: str) -> None:
+        self.client = Pinecone(api_key=api_key)
+
+    def initialize_index(
```
Comment (orhanrauf) [inline backend/app/core/task_definition_retriever.py:29]: Fine for now, but perhaps best moved to Factory class later when requirements for creating indexes get multifaceted and the PineconeService gets too heavy

File: backend/app/core/task_definition_retriever.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    def __init__(self, api_key: str = None) -> None:
+        self.client = OpenAI(api_key=api_key)
+
+    def get_embedding(self, text: str, model="text-embedding-3-small") -> list[float]:
+        text = text.replace("\n", " ")
+        response = self.client.embeddings.create(input=[text], model=model)
+        return response.data[0].embedding
+
+
+class TaskDefinitionRetrievalManager:
```
Comment (orhanrauf) [inline backend/app/core/task_definition_retriever.py:103]: Curious why RetrievalManager naming was opted for instead of Retriever for example

**Review by orhanrauf — APPROVED** [review]
Looks great, just some design remarks. Let's make mental notes and proceed

File: backend/app/core/task_definition_retriever.py (hunk 2 above)
Comment (lennertjansen) [reply backend/app/core/task_definition_retriever.py:73]: So I should remove the line `logger.error(f"Timeout occurred while initializing the Pinecone index '{index_name}'.", exc_info=True)` and just raise the `exception`?

File: backend/app/core/task_definition_retriever.py (hunk 4 above)
Comment (lennertjansen) [reply backend/app/core/task_definition_retriever.py:103]: To more accurately reflect the class' (somewhat bloated) responsibilities. It does more than just retrieval; it also manages the gathering of all task definitions from the relational database, and the population of the vector index.

That being said. Future refactoring should involve breaking up this class into more specialized classes with fewer responsibilities.

File: backend/app/core/task_definition_retriever.py (hunk 3 above)
Comment (lennertjansen) [reply backend/app/core/task_definition_retriever.py:29]: You're referring to moving the `initialize_index()` method to a Factory class, right? I'm asking so I can move this to a new issue for later.

Comment (orhanrauf) [reply backend/app/core/task_definition_retriever.py:29]: Yes sounds good: https://github.com/users/orhanrauf/projects/3/views/1?pane=issue&itemId=56015354

**Pull request merged**